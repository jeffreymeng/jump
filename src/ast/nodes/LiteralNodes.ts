import ASTNode from "./ASTNode";
import { JumpInternalError } from "../JumpInternalError";

abstract class LiteralNode<T> extends ASTNode<T> {
	public readonly value: T;

	constructor(symbol: T) {
		super();
		this.value = symbol;
	}
}

export class DoubleNode extends LiteralNode<number> {
	constructor(symbol: string) {
		const value = parseFloat(symbol);
		super(value);
	}

	public evaluate() {
		return this.value;
	}

	public toJSON() {
		return {
			name: "DoubleNode",
			value: this.value,
		};
	}
}

export class IntNode extends LiteralNode<number> {
	constructor(symbol: string) {
		const value = parseFloat(symbol);
		if (value % 1 !== 0) {
			throw new JumpInternalError(
				"IntNode created with non-integer value."
			);
		}
		super(value);
	}

	public evaluate() {
		return this.value;
	}

	public toJSON() {
		return {
			name: "IntNode",
			value: this.value,
		};
	}
}

export class StringNode extends LiteralNode<string> {
	public evaluate() {
		return `"${this.value.replace(/"/g, '\\"')}"`;
	}

	public toJSON() {
		return {
			name: "StringNode",
			value: this.value,
		};
	}
}
