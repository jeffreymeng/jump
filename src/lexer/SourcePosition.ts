/**
 * A SourcePosition refers to a specific position in the source.
 * Each character in the source has a unique position.
 */
export default class SourcePosition {
	/**
	 * @param source - The text of the source code.
	 * @param index - The index of the character within the source text.
	 */
	constructor(protected readonly source: string, protected readonly index: number) {
		if (index >= source.length) {
			throw new Error("SourcePosition expected an index less than the length of source.")
		}
	}

	/**
	 * Returns the line of the index in the given source text. Lines are 1-indexed.
	 */
	public getLine(): number {
		return this.source.substring(0, this.index).split("\n").length
	}

	/**
	 * Returns the index of the character within its line. Columns are
	 * 1-indexed.
	 */
	public getCol(): number {
		// given the column, look backwards for the closest newline strictly
		// before the current index, since if the current character is a
		// newline that doesn't count.
		return this.index - this.source.lastIndexOf("\n", this.index - 1);
	}

}
