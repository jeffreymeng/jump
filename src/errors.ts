import SourcePosition from "./lexer/SourcePosition";

export class JumpError extends Error {
	constructor(message: string) {
		super(message);
	}
}

/**
 * An error in the syntax of the input.
 */
export class JumpSyntaxError extends JumpError {
	public readonly position: SourcePosition;

	constructor(message: string, position: SourcePosition) {
		super(message);
		this.position = position;
	}
}

export class JumpInternalError extends JumpError {
	constructor(message: string) {
		super(message);
	}
}

export class JumpDivisionByZeroError extends JumpError {
	constructor(dividend: number, public readonly position: SourcePosition) {
		super(`JumpDivisionByZeroError: Cannot divide ${dividend} by zero.`);
	}
}

/**
 * An error in the implementation of Jump's AST.
 */
export class JumpASTInternalError extends JumpError {
	constructor(message: string) {
		super(message);
	}
}
