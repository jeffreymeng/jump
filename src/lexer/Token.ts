import SourcePosition from "./SourcePosition";

export enum TokenType {
	STRING_LITERAL = "string_literal",
	INT_LITERAL = "int_literal",
	DOUBLE_LITERAL = "double_literal",
	OPERATOR = "operator",
	IDENTIFIER = "identifier",
	CONTROL = "control",
	KEYWORD = "keyword",
}
export default class Token {
	/**
	 *
	 * @param type - The type of the token
	 * @param symbol - The value of the token
	 */
	constructor(
		public readonly type: TokenType,
		public readonly symbol: string,
		public readonly position: SourcePosition
	) {}

	/**
	 * Returns whether or not the token matches another token, in terms of content,
	 * but not position.
	 * @param other
	 */
	public equals(other: Token) {
		return this.type === other.type && this.symbol === other.symbol;
	}

	/**
	 * Returns whether or not the token matches the provided type and, if provided, symbol.
	 * @param type
	 * @param symbol
	 */
	public is(type: TokenType, symbol?: string) {
		return this.type === type && (!symbol || this.symbol === symbol);
	}

	public toString() {
		return `Token<${this.type}, '${this.symbol}'>`;
	}
}
