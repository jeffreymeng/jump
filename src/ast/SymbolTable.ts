import SourcePosition from "../lexer/SourcePosition";
import { JumpInternalError } from "../errors";

export enum SCOPE_TYPE {
	LOOP = "loop",
	FUNCTION = "function",
	GLOBAL = "global",
}

/**
 * Returns a new array with each element in reverse order.
 * @param arr - The array to reverse. This array will not be modified.
 */
function reversed<T>(arr: T[]): T[] {
	const rev = arr.slice();
	rev.reverse();
	return rev;
}

export default class SymbolTable {
	protected table: {
		symbols: Map<
			string,
			{
				type: string;
				value: any;
			}
		>;
		name: string;
		type: SCOPE_TYPE;
		position: SourcePosition;
	}[] = [
		{
			name: "Global",
			type: SCOPE_TYPE.GLOBAL,
			position: new SourcePosition("", 0),
			symbols: new Map(),
		},
	];

	/**
	 * Creates a new scope to add to the table.
	 * @param name - The name of the scope.
	 * @param type - The type of the scope.
	 * @param position - The position of the start of the scope in the source.
	 */
	public pushScope(
		name: string,
		type: SCOPE_TYPE,
		position: SourcePosition
	): void {
		// name: 'Loop 147 @ L13:4'
		// name: 'parse() @ L19:2'
		this.table.push({
			name,
			type,
			position,
			symbols: new Map(),
		});
	}

	/**
	 * Exits the topmost scope of the stack, removing any values stored in that scope.
	 */
	public exitScope(): void {
		if (this.table.length <= 1) {
			throw new JumpInternalError("Cannot exit the global scope!");
		}
		this.table.pop();
	}

	/**
	 * Check if a symbol with a given identifier exists.
	 * @param id - The identifier of the variable.
	 */
	public has(id: string) {
		return reversed(this.table).some((table) => table.symbols.has(id));
	}

	/**
	 * Declare a variable. The variable must not already exist.
	 * @param id - The identifier of the variable.
	 * @param type - The type of the variable. Types are not checked in the symbol table.
	 * @param value - The value of the variable.
	 */
	public declare(id: string, type: string, value: any): void {
		if (this.table.at(-1)!.symbols.has(id)) {
			throw new JumpInternalError(
				`Identifier '${id}' has already been declared.`
			);
		}
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		this.table.at(-1)!.symbols.set(id, {
			type,
			value,
		});
	}

	/**
	 * Update a variable. The variable must already exist.
	 * @param id - The identifier of the variable.
	 * @param value - The new value of the variable.
	 */
	public update(id: string, val: any) {
		for (const table of reversed(this.table)) {
			if (table.symbols.has(id)) {
				table.symbols.set(id, {
					...table.symbols.get(id)!,
					value: val,
				});
				return;
			}
		}
		throw new JumpInternalError(
			`Unable to update identifier '${id}' because it was not found in the symbol table.`
		);
	}

	/**
	 * Gets the type and value of a variable. The variable must already exist.
	 * @param id - The identifier of the variable.
	 */
	public get(id: string): { type: string; value: any } {
		for (const table of reversed(this.table)) {
			if (table.symbols.has(id)) {
				return table.symbols.get(id)!;
			}
		}
		throw new JumpInternalError(
			`Unable to get identifier '${id}' because it was not found in the symbol table.`
		);
	}
}
