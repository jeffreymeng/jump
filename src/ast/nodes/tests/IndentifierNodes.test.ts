import { describe, expect, test, jest, beforeAll } from "@jest/globals";
import {
	Identifier,
	TypeIdentifier,
	VariableAssignmentNode,
	VariableDeclarationNode,
	VariableNode,
} from "../IdentifierNodes";
import Token, { TokenType } from "../../../lexer/Token";
import SourcePosition from "../../../lexer/SourcePosition";
import { JumpInternalError } from "../../../errors";
import { IntNode } from "../LiteralNodes";
import SymbolTable from "../../SymbolTable";

jest.mock("../../SymbolTable");
const mockSymbolTable = jest.mocked(SymbolTable);

describe("Identifiers", () => {
	test("Basic identifier", () => {
		const i = new Identifier("foo");
		expect(i.name).toBe("foo");
	});
	test("Basic identifier from token", () => {
		const sp = new SourcePosition("", 0);
		const i = Identifier.fromToken(
			new Token(TokenType.IDENTIFIER, "bar", sp)
		);
		const i2 = Identifier.fromToken(
			new Token(TokenType.IDENTIFIER, "bar", sp)
		);
		expect(i).not.toBe(i2);
		expect(i).toStrictEqual(i2);
		expect(i.name).toBe("bar");
		expect(() =>
			Identifier.fromToken(new Token(TokenType.OPERATOR, "+", sp))
		).toThrow(JumpInternalError);
	});
});

describe("VariableDeclarationNode", () => {
	const node = new VariableDeclarationNode(
		new Identifier("foo"),
		new TypeIdentifier("bar"),
		new IntNode("3")
	);

	beforeAll(() => {
		mockSymbolTable.mockClear();
	});

	test("Declares a variable on the symbol table when evaluated", () => {
		node.evaluate(new SymbolTable());
		expect(mockSymbolTable.mock.instances[0].declare).toHaveBeenCalledWith(
			"foo",
			"bar",
			3
		);
	});

	test("Creates the same json when toJSON() is called", () => {
		expect(node.toJSON()).toMatchSnapshot();
	});
});

describe("VariableDeclarationNode", () => {
	const node = new VariableAssignmentNode(
		new Identifier("foo"),
		new IntNode("3")
	);

	beforeAll(() => {
		mockSymbolTable.mockClear();
	});

	test("Updates a variable on the symbol table when evaluated", () => {
		node.evaluate(new SymbolTable());
		expect(mockSymbolTable.mock.instances[0].update).toHaveBeenCalledWith(
			"foo",
			3
		);
	});

	test("Creates the same json when toJSON() is called", () => {
		expect(node.toJSON()).toMatchSnapshot();
	});
});

describe("VariableNode", () => {
	const node = new VariableNode(new Identifier("foo"));

	beforeAll(() => {
		mockSymbolTable.mockClear();
	});

	// TODO
	// test("Retrieves a variable from the symbol table when evaluated", () => {
	// 	mockSymbolTable.mockImplementation(() => ({
	// 		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
	// 		// @ts-ignore
	// 		get: (id) => ({
	// 			type: "int",
	// 			value: 10,
	// 		}),
	// 	}));
	// 	expect(node.evaluate(new SymbolTable())).toBe(10);
	// 	expect(mockSymbolTable.mock.instances[0].get).toHaveBeenCalledWith(
	// 		"foo"
	// 	);
	// });

	test("Creates the same json when toJSON() is called", () => {
		expect(node.toJSON()).toMatchSnapshot();
	});
});
