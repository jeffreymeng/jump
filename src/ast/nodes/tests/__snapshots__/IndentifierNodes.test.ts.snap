// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeclarationNode Creates the same json when toJSON() is called 1`] = `
{
  "node": "Declaration: type [object Object], id [object Object]",
  "value": {
    "node": "Int Literal: 3",
  },
}
`;

exports[`VariableDeclarationNode Creates the same json when toJSON() is called 1`] = `
{
  "node": "Declaration: type [object Object], id [object Object]",
  "value": {
    "node": "Int Literal: 3",
  },
}
`;

exports[`VariableDeclarationNode Creates the same json when toJSON() is called 2`] = `
{
  "node": "Assignment: [object Object]",
  "value": {
    "node": "Int Literal: 3",
  },
}
`;

exports[`VariableNode Creates the same json when toJSON() is called 1`] = `
{
  "node": "Variable: id [object Object]",
}
`;
